flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/GatedRateLimiter.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.connector.source.lib.util;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * An implementation of {@link RateLimiter} that completes defined number of futures in-between the
+ * external notification events. The first cycle completes immediately, without waiting for the
+ * external notifications.
+ */
+public class GatedRateLimiter implements RateLimiter {
+
+    private final int capacityPerCycle;
+    private int capacityLeft;
+
+    /**
+     * Instantiates a new GatedRateLimiter.
+     *
+     * @param capacityPerCycle The number of completed futures per cycle.
+     */
+    public GatedRateLimiter(int capacityPerCycle) {
+        this.capacityPerCycle = capacityPerCycle;
+        this.capacityLeft = capacityPerCycle + 1;
+    }
+
+    CompletableFuture<Void> gatingFuture;
+
+    @Override
+    public CompletionStage<Void> acquire() {
+        if (capacityLeft-- > 0) {
+            return CompletableFuture.completedFuture(null);
+        } else {
+            if (gatingFuture == null) {
+                gatingFuture = new CompletableFuture<>();
+            }
+            return gatingFuture;
+        }
+    }
+
+    @Override
+    public void notifyCheckpointComplete(long checkpointId) {
+        capacityLeft = capacityPerCycle - 1;
```

Comment:
Let me try to clarify what is going on:
- The goal is to always emit "capacity" values per cycle.
- The first cycle after the start emits the "capacity" number the elements immediately, without waiting for the checkpoint. This is why the constructor initially gives capacity one more element - the future that is normally completed in the notifyCheckpointComplete has to be completed immediately in the first run.
- Subsequent cycles produce the same number of elements but all of them are only released when the gating future completes. In `notifyCheckpointComplete` the new cycle starts. The gating future completes, causing the release of the first element of the new cycle; since the first result has just been emitted by the gating future, the remaining capacity available to the acquire method needs to be reduced by one.
The gating future basically always "belongs" to the new cycle, triggered by the checkpoint. It emits an element, hence -1. Since in the first cycle we do not have the gating future, we add +1 to "emulate" it. Hope it makes sense. You can check that it works correctly here: [DataGeneratorPerCheckpoint](https://github.com/apache/flink/blob/608a43d28e0f5c0fef097b6581d8bb43ce983605/flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/datagen/DataGeneratorPerCheckpoint.java#L36)

Reply:
I think it's a problem that the first cycle is somehow treated differently than the others; conceptually it should be _possible_ to just call notifyCheckpointComplete immediately (e.g., in the constructor) without it affecting the behavior.

AFAICT the reason we need this distinction is that we decrement the count _before_ the future completes, causing us to smuggle some decrements over.

I'm currently experimenting to avoid that problem (and others incidentally):

```
    CompletableFuture<Void> gatingFuture = CompletableFuture.completedFuture(null);

    @Override
    public CompletionStage<Void> acquire() {
        if (capacityLeft <= 0) {
            gatingFuture = new CompletableFuture<>();
        }
        return gatingFuture.thenRun(() -> capacityLeft -= 1);
    }

    @Override
    public void notifyCheckpointComplete(long checkpointId) {
        capacityLeft = capacityPerCycle;
        gatingFuture.complete(null);
    }
```

Reply:
Without having played around with the `DataGeneratorPerCheckpoint` it doesn't quite make sense to me. Why would we only emit 9 values in the first cycle? It doesn't match the results from the unit test I proposed below, that shows that my version emits exactly `capacity` values per cycle.

Reply:
While debugging I noticed that acquire() is called twice initially without the first value being printed. This could indicate a bug in the source somewhere.

Reply:
The problem is that `IteratorSourceReaderBase#pollNext` does not emit a value if `iterator == null` and we moved to the next split. However the we null the `availabilityFuture` in `RateLImitedSourceReader#pollNext` anyway.

So we're just blow 1 capacity in the first cycle.

Reply:
amend `IteratorSourceReaderBase#pollNext` and follows to fix it:

```
-return tryMoveToNextSplit();
+final InputStatus inputStatus = tryMoveToNextSplit();
+if (inputStatus == InputStatus.MORE_AVAILABLE) {
+    output.collect(convert(iterator.next()));
+}
+return inputStatus;
```

Reply:
I guess a similar issue would currently occur if the split is complete.

Reply:
Aha, so that helped to surface an actual issue. Neat. Added the fix and the respective test. 
Since the lifecycle of the reader is bound to the lifecycle of the split/iterator, I do not think that split completion should be an issue.